src/nnet3/nnet-simple-component.cc

```diff
@@ -1011,6 +1022,196 @@ void PerElementScaleComponent::UnVectorize(
+}
+
+void PerElementOffsetComponent::Init(
+    BaseFloat learning_rate, int32 dim,
+    BaseFloat param_mean, BaseFloat param_stddev) {
+  UpdatableComponent::Init(learning_rate);
+  offsets_.Resize(dim);
+  KALDI_ASSERT(dim > 0 && param_stddev >= 0.0);
+  offsets_.SetRandn();
+  offsets_.Scale(param_stddev);
+  offsets_.Add(param_mean);
+}
+
+void PerElementOffsetComponent::Init(BaseFloat learning_rate,
+                                    std::string matrix_filename) {
+  UpdatableComponent::Init(learning_rate);
+  CuMatrix<BaseFloat> mat;
+  ReadKaldiObject(matrix_filename, &mat); // will abort on failure.
+  KALDI_ASSERT(mat.NumCols() == 1);
+  int32 dim = mat.NumRows();
+  offsets_.Resize(dim);
+  offsets_.CopyColFromMat(mat, 0);
+}
+
+void PerElementOffsetComponent::InitFromConfig(ConfigLine *cfl) {
+  bool ok = true;
+  BaseFloat learning_rate = learning_rate_;
+  std::string matrix_filename;
+  int32 dim = -1;
+  cfl->GetValue("learning-rate", &learning_rate); // optional.
+  if (cfl->GetValue("matrix", &matrix_filename)) {
+    Init(learning_rate, matrix_filename);
+    if (cfl->GetValue("dim", &dim))
+      KALDI_ASSERT(dim == InputDim() &&
+                   "input-dim mismatch vs. matrix.");
+  } else {
+    ok = ok && cfl->GetValue("dim", &dim);
+    BaseFloat param_mean = 1.0, param_stddev = 0.0;
+    cfl->GetValue("param-mean", &param_mean);
+    cfl->GetValue("param-stddev", &param_stddev);
+    Init(learning_rate, dim, param_mean, param_stddev);
+  }
+  if (cfl->HasUnusedValues())
+    KALDI_ERR << "Could not process these elements in initializer: "
+              << cfl->UnusedValues();
+  if (!ok)
+    KALDI_ERR << "Bad initializer " << cfl->WholeLine();
+}
+
+void PerElementOffsetComponent::Propagate(
+    const ComponentPrecomputedIndexes *indexes,
+    const CuMatrixBase<BaseFloat> &in,
+    CuMatrixBase<BaseFloat> *out) const {
+  
+  out->CopyRowsFromVec(offsets_);
+  out->AddMat(1.0, in, kNoTrans);
+}
+
+void PerElementOffsetComponent::UpdateSimple(
+    const CuMatrixBase<BaseFloat> &in_value,
+    const CuMatrixBase<BaseFloat> &out_deriv) {
+
+  offsets_.AddRowSumMat(learning_rate_, out_deriv, 1.0);
+}
+
+void PerElementOffsetComponent::Backprop(
+    const std::string &debug_info,
+    const ComponentPrecomputedIndexes *indexes,
+    const CuMatrixBase<BaseFloat> &in_value,
+    const CuMatrixBase<BaseFloat> &, // out_value
+    const CuMatrixBase<BaseFloat> &out_deriv,
+    Component *to_update_in,
+    CuMatrixBase<BaseFloat> *in_deriv) const {
+  PerElementOffsetComponent *to_update =
+      dynamic_cast<PerElementOffsetComponent*>(to_update_in);
+
+  if (in_deriv) {
+    // Propagate the derivative back to the input.
+    in_deriv->CopyFromMat(out_deriv);
+  }
+
+  if (to_update != NULL) {
+    // Next update the model (must do this 2nd so the derivatives we propagate
+    // are accurate, in case this == to_update_in.)
+    if (to_update->is_gradient_)
+      to_update->UpdateSimple(in_value, out_deriv);
```

Comment:
There would be no UpdateSimple function- you should rewrite the function to
do any update in-place.  If you were to keep UpdateSimple (and don't do
this), then you would remove the extra argument.
For the Propagate and Backprop functions, you generally pass an empty
matrix for unused inputs.
Dan


On Mon, Nov 2, 2015 at 12:53 AM, hhadian <notifications@github.com> wrote:

> In src/nnet3/nnet-simple-component.cc
> <https://github.com/kaldi-asr/kaldi/pull/307#discussion_r43598840>:
>
> > +    const CuMatrixBase<BaseFloat> &out_deriv,
> > +    Component *to_update_in,
> > +    CuMatrixBase<BaseFloat> *in_deriv) const {
> > +  PerElementOffsetComponent *to_update =
> > +      dynamic_cast<PerElementOffsetComponent*>(to_update_in);
> > +
> > +  if (in_deriv) {
> > +    // Propagate the derivative back to the input.
> > +    in_deriv->CopyFromMat(out_deriv);
> > +  }
> > +
> > +  if (to_update != NULL) {
> > +    // Next update the model (must do this 2nd so the derivatives we propagate
> > +    // are accurate, in case this == to_update_in.)
> > +    if (to_update->is_gradient_)
> > +      to_update->UpdateSimple(in_value, out_deriv);
>
> What to pass in place of in_value? I mean wrt kaldi coding standards. Is
> UpdateSimple(out_deriv, out_deriv) ok?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/307/files#r43598840>.
>